### PR TITLE
Added password validation check within myaccount change password section

### PR DESCRIPTION
**Pull request by gershom96**
### Purpose
The change password section in myaccount did not have a password validation section. This was added in this PR.

### Checklist
- [ ] e2e cypress tests locally verified.
- [ ] Manual test round performed and verified.
- [ ] UX/UI review done on the final implementation.
- [ ] Documentation provided. (Add links if there are any)
- [ ] Unit tests provided. (Add links if there are any)
- [ ] Integration tests provided. (Add links if there are any)

### Security checks
- [ ] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines?
- [ ] Ran FindSecurityBugs plugin and verified report?
- [ ] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?

### CODE REVIEW

File: apps/myaccount/src/components/change-password/change-password.tsx [hunk 1]
```diff
@@ -80,6 +81,67 @@ export const ChangePassword: FunctionComponent<ChangePasswordProps> = (props: Ch
 
     const endUserSession = useEndUserSession();
 
+    interface iconProps {
+        class : string;
```
Comment (yasinmiran) [inline apps/myaccount/src/components/change-password/change-password.tsx:85]: Avoid using `class` or any keywords in attribute namespaces.

```suggestion
        cssClassName : string;
```

File: apps/myaccount/src/components/change-password/change-password.tsx [hunk 2]
```diff
@@ -80,6 +81,67 @@ export const ChangePassword: FunctionComponent<ChangePasswordProps> = (props: Ch
...
+            color : "grey" as SemanticCOLORS
+        };
+
+        const lowerCaseLetters = /[a-z]/g;
+        const upperCaseLetters = /[A-Z]/g;
+        const numbers = /[0-9]/g;
+        const chars = /[!@#$%^&*~]/g;
+
+        if ( id === "password-validation-length" ){
+            if (password === "") {
```
Comment (yasinmiran) [inline apps/myaccount/src/components/change-password/change-password.tsx:102]: Use a constant

const EMPTY_STRING = "";

```suggestion
            if (password === EMPTY_STRING) {
```

File: apps/myaccount/src/components/change-password/change-password.tsx [hunk 3]
```diff
@@ -80,6 +81,67 @@ export const ChangePassword: FunctionComponent<ChangePasswordProps> = (props: Ch
...
+                iconProps.class = iconProps.class + " thin"
+            } else if ( password.length > 8 ) {
+                iconProps.class = "check " + iconProps.class
+                iconProps.color = "green" as SemanticCOLORS;
+            } else {
+                iconProps.class = "remove " + iconProps.class
+                iconProps.color = "red" as SemanticCOLORS;
+            }
+        } else if ( id === "password-validation-case" ) {
```
Comment (yasinmiran) [inline apps/myaccount/src/components/change-password/change-password.tsx:111]: No need to keep spaces inside before/after the parenthesis.
Let's use ESLint plugin to format the code.

```suggestion
        } else if (id === "password-validation-case") {
```

File: apps/myaccount/src/components/change-password/change-password.tsx [hunk 4]
```diff
@@ -315,6 +377,31 @@ export const ChangePassword: FunctionComponent<ChangePasswordProps> = (props: Ch
                             />
                         </Suspense>
                     </Form.Field>
+                    
+                    { passwordValidationConfig.showPasswordValidation  &&
+                        <Form.Field width={ 9 } data-testid={ `${testId}-new-password-validation-field` }>
+                            <div className="password-policy-description">                                
+                                <Icon id = "password-validation-length" className = {getIconProps("password-validation-length").class} color = {getIconProps("password-validation-length").color}/>
```
Comment (yasinmiran) [inline apps/myaccount/src/components/change-password/change-password.tsx:384]: No spaces before/after equal when passing down props. Please use ESLint.

```suggestion
                                <Icon id="password-validation-length" className={getIconProps("password-validation-length").class} color={getIconProps("password-validation-length").color}/>
```

File: apps/myaccount/src/components/change-password/change-password.tsx [hunk 5]
```diff
@@ -315,6 +377,31 @@ export const ChangePassword: FunctionComponent<ChangePasswordProps> = (props: Ch
...
+                                <p>More than 8 characters</p>
+                            </div>
+                            
+                            <div className="password-policy-description">                                
+                                <Icon id = "password-validation-case" className = {getIconProps("password-validation-case").class} color = {getIconProps("password-validation-case").color} />
+                                <p>At least one uppercase and lowercase letter</p>
```
Comment (yasinmiran) [inline apps/myaccount/src/components/change-password/change-password.tsx:390]: Shall we declare and get these strings from `i18n` module?

```suggestion
                                <p>At least one uppercase and lowercase letter</p>
```

File: apps/myaccount/src/components/change-password/change-password.tsx [hunk 6]
```diff
@@ -315,6 +377,31 @@ export const ChangePassword: FunctionComponent<ChangePasswordProps> = (props: Ch
...
+                            </div>
+                            
+                            <div className="password-policy-description">                                
+                                <Icon id = "password-validation-number" className = {getIconProps("password-validation-number").class} color = {getIconProps("password-validation-number").color}/>
```
Comment (yasinmiran) [inline apps/myaccount/src/components/change-password/change-password.tsx:394]: These lines are too long. Format props into multilines. Use ESLint.

```suggestion
                                <Icon id = "password-validation-number" className = {getIconProps("password-validation-number").class} color = {getIconProps("password-validation-number").color}/>
```

File: apps/myaccount/src/components/change-password/change-password.tsx [hunk 7]
```diff
@@ -80,6 +81,67 @@ export const ChangePassword: FunctionComponent<ChangePasswordProps> = (props: Ch
...
+        const chars = /[!@#$%^&*~]/g;
+
+        if ( id === "password-validation-length" ){
+            if (password === "") {
+                iconProps.class = iconProps.class + " thin"
+            } else if ( password.length > 8 ) {
+                iconProps.class = "check " + iconProps.class
+                iconProps.color = "green" as SemanticCOLORS;
+            } else {
+                iconProps.class = "remove " + iconProps.class
```
Comment (yasinmiran) [inline apps/myaccount/src/components/change-password/change-password.tsx:108]: These conditions are bloated. You can simplify this into: -

```
const DEFAULT: IconProps = {
    cssClassName: "circle",
    color: "grey"
};

const POSITIVE: IconProps = {
    cssClassName: `${DEFAULT.cssClassName} check`,
    color: `green`
};

const NEGATIVE: IconProps = { ... }

if (password === "") {
    return DEFAULT;
} else if ( password.match(chars) ) {
    return POSITIVE;
} else {
    return NEGATIVE;
}
```

**Review by yasinmiran — CHANGES_REQUESTED** [review]

File: apps/myaccount/src/components/change-password/change-password.tsx [hunk 8]
```diff
@@ -79,6 +90,64 @@ export const ChangePassword: FunctionComponent<ChangePasswordProps> = (props: Ch
     const dispatch = useDispatch();
 
     const endUserSession = useEndUserSession();
+    const EMPTY_STRING = "";
+
+    const getIconProps = (id : string) : IconProps =>{
```
Comment (DimalChandrasiri) [inline apps/myaccount/src/components/change-password/change-password.tsx:95]: Add a doc comment regarding the implementation and usage

File: apps/myaccount/src/extensions/configs/models/password-validation.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+export interface PasswordValidationConfig {
```
Comment (DimalChandrasiri) [inline apps/myaccount/src/extensions/configs/models/password-validation.ts:19]: Update the license header to new one.

File: apps/myaccount/src/extensions/configs/password-validation.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { PasswordValidationConfig } from "./models/password-validation";
```
Comment (DimalChandrasiri) [inline apps/myaccount/src/extensions/configs/password-validation.ts:19]: Update the license header to new one

**Review by DimalChandrasiri — APPROVED** [review]

**Review by yasinmiran — APPROVED** [review]

File: apps/myaccount/src/components/change-password/change-password.tsx [hunk 11]
```diff
@@ -315,6 +388,47 @@ export const ChangePassword: FunctionComponent<ChangePasswordProps> = (props: Ch
                             />
                         </Suspense>
                     </Form.Field>
+                    
+                    { passwordValidationConfig.showPasswordValidation  &&
```
Comment (gershom96) [inline apps/myaccount/src/components/change-password/change-password.tsx:392]: Add an enum for the id strings.

Comment (gershom96) [inline apps/myaccount/src/components/change-password/change-password.tsx:392]: paranthesis preceding the form should start right after the &&

Comment (wenodpathirana) [inline apps/myaccount/src/components/change-password/change-password.tsx:392]: Additional space before `&&`

File: modules/i18n/src/translations/en-US/portals/myaccount.ts [hunk 12]
```diff
@@ -263,6 +263,7 @@ export const myAccount: MyAccountNS = {
                         }
                     },
                     validations: {
+
```
Comment (wenodpathirana) [inline modules/i18n/src/translations/en-US/portals/myaccount.ts:266]: remove the extra line.

**Pull request merged**